flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java

```diff
@@ -53,283 +49,111 @@
 	@Override
 	public void run() {
 		try {
-			String nextShardItr = getShardIterator(lastSequenceNum);
+			while (isRunning()) {
+				final RecordPublisherRunResult result = recordPublisher.run(startingPosition, batch -> {
+					for (UserRecord userRecord : batch.getDeaggregatedRecords()) {
+						if (filterDeaggregatedRecord(userRecord)) {
+							deserializeRecordForCollectionAndUpdateState(userRecord);
+						}
+					}
 
-			long processingStartTimeNanos = System.nanoTime();
+					shardConsumerMetricsReporter.setAverageRecordSizeBytes(batch.getAverageRecordSizeBytes());
+					shardConsumerMetricsReporter.setNumberOfAggregatedRecords(batch.getAggregatedRecordSize());
+					shardConsumerMetricsReporter.setNumberOfDeaggregatedRecords(batch.getDeaggregatedRecordSize());
+					ofNullable(batch.getMillisBehindLatest()).ifPresent(shardConsumerMetricsReporter::setMillisBehindLatest);
+				});
 
-			while (isRunning()) {
-				if (nextShardItr == null) {
+				if (result == COMPLETE) {
 					fetcherRef.updateState(subscribedShardStateIndex, SentinelSequenceNumber.SENTINEL_SHARD_ENDING_SEQUENCE_NUM.get());
 
 					// we can close this consumer thread once we've reached the end of the subscribed shard
 					break;
 				} else {
-					shardMetricsReporter.setMaxNumberOfRecordsPerFetch(maxNumberOfRecordsPerFetch);
-					GetRecordsResult getRecordsResult = getRecords(nextShardItr, maxNumberOfRecordsPerFetch);
-
-					List<Record> aggregatedRecords = getRecordsResult.getRecords();
-					int numberOfAggregatedRecords = aggregatedRecords.size();
-					shardMetricsReporter.setNumberOfAggregatedRecords(numberOfAggregatedRecords);
-
-					// each of the Kinesis records may be aggregated, so we must deaggregate them before proceeding
-					List<UserRecord> fetchedRecords = deaggregateRecords(
-						aggregatedRecords,
-						subscribedShard.getShard().getHashKeyRange().getStartingHashKey(),
-						subscribedShard.getShard().getHashKeyRange().getEndingHashKey());
-
-					long recordBatchSizeBytes = 0L;
-					for (UserRecord record : fetchedRecords) {
-						recordBatchSizeBytes += record.getData().remaining();
-						deserializeRecordForCollectionAndUpdateState(record);
-					}
-
-					int numberOfDeaggregatedRecords = fetchedRecords.size();
-					shardMetricsReporter.setNumberOfDeaggregatedRecords(numberOfDeaggregatedRecords);
-
-					nextShardItr = getRecordsResult.getNextShardIterator();
-
-					long adjustmentEndTimeNanos = adjustRunLoopFrequency(processingStartTimeNanos, System.nanoTime());
-					long runLoopTimeNanos = adjustmentEndTimeNanos - processingStartTimeNanos;
-					maxNumberOfRecordsPerFetch = adaptRecordsToRead(runLoopTimeNanos, fetchedRecords.size(), recordBatchSizeBytes, maxNumberOfRecordsPerFetch);
-					shardMetricsReporter.setRunLoopTimeNanos(runLoopTimeNanos);
-					processingStartTimeNanos = adjustmentEndTimeNanos; // for next time through the loop
+					startingPosition = StartingPosition.continueFromSequenceNumber(lastSequenceNum);
```

Comment:
Why does the `ShardConsumer` need to keep track and update the `startingPosition`?
It feels like this should be an internal state to the record publisher, and also as how I understand it, the record publisher only ever respects the `nextShardItr` that it maintains.

Reply:
See other comment regarding `StartingPosition`

Reply:
It does not anymore.